### ISSUE TITLE
"High Hopes" Null Dereference pt. 2

### ISSUE THREAD

**Issue by turnerrocks1** [blacktop/ipsw#62]

**Comment by blacktop** [comment]
LOL 🤣

**Comment by turnerrocks1** [comment]
Ok so this time around unlike last time after wrapping my head around your last commit and actually getting a feel for go language it's safe to assume that your code now actually parses the crash log correctly this time around the only thing is now that it's not finding the image in the shared cache file. I had the same issue when reading the shared cache from a javascript exploit but it was easily fixable 
If you don't mind me dropping my implementation of the parser in javascript language here and maybe you can translate it to corresponding go language or maybe see what needs to be added as my parser was meant for this "raw" dyld_shared_cache of "old" Macho Type.

Reference : https://github.com/turnerrocks1/turnerrocks1.github.io/blob/master/index1.html
"       //log('[*] We will now walk down the code to find the illegal instruction for demo purposes...');
        //tmp = base;
        //var hdr = Add(0x180000000,slide);
       // log(hdr + hdr1)
        while(true)
        {
        //if (memory.readInt64(hdr).lo() == 4277009104) {
          //alert('WebCore ' + hdr + ' post spectre support coming soon...mightnotwork');
        //}
        if(strcmp(memory.read(hdr, 0x10), "dyld_v1   arm64"))
        {
            //alert(memory.read(hdr, 0x10)[0]))
            break;
        }
        hdr = Sub(hdr, 0x1000);
        }
        //alert('dyld cache header @' + hdr);
        var uuid = memory.u32(Add(hdr, 0x58));
       // alert("uuid:" + uuid) //+ "uuid..."+ memory.read_i64(hdr + 0x58))
        //var dcachedump = String.fromCharCode(...memory.copyfrom(hdr,0,100))
        
        function ab2str(buf) {
  return String.fromCharCode.apply(null, new Uint16Array(buf));
}
        //alert('Mach-O magic' + memory.u32(hdr))
        var base_seg = null;
       // alert("nsegs test" + memory.read_i32(hdr+0x14))
        var nsegs    = memory.u32(Add(hdr, 0x14));
       // alert(memory.u32(hdr + 0x18) + "segment"+ nsegs*0x20 + "sets")
        var segdata  = memory.read(Add(hdr, memory.u32(Add(hdr, 0x10))), nsegs * 0x20);
        var segs     = [];
        for(var z = 0; z < nsegs; ++z)
        {
            var off = z * 0x20;
            var seg =
            {
            addr:     new Int64(segdata.slice(off +  0x0, off +  0x8)),
            size:     new Int64(segdata.slice(off +  0x8, off + 0x10)),
            fileoff:  new Int64(segdata.slice(off + 0x10, off + 0x18)),
            maxprot:  b2u32(segdata.slice(off + 0x18, off + 0x1c)),
            initprot: b2u32(segdata.slice(off + 0x1c, off + 0x20))
            };
            segs.push(seg);
            var max_rights = "";
            var init_rights = "";
            if(seg.maxprot & VM_PROT_NONE){
                max_rights += "-";
            }
            if(seg.maxprot & VM_PROT_READ) {
                max_rights += "r";
            }
            if(seg.maxprot & VM_PROT_WRITE) {
                max_rights += "w";
            }
            if(seg.maxprot & VM_PROT_EXECUTE){
                max_rights += "x";
            }
            if(seg.initprot & VM_PROT_NONE){
                init_rights += "-";
            }
            if(seg.initprot & VM_PROT_READ) {
                init_rights += "r";
            }
            if(seg.initprot & VM_PROT_WRITE) {
                init_rights += "w";
            }
            if(seg.initprot & VM_PROT_EXECUTE){
                init_rights += "x";
            }
            if(max_rights.length < 4){
                max_rights = "-" + max_rights;
            }
            if(init_rights.length < 4){
                init_rights = "-" + init_rights;
            }
           // log("DYLD Shared cache segment ("+seg.addr.toString(16)+"-"+Add(seg.addr, seg.size).toString(16)+"): "+init_rights+"/"+max_rights);
            //memory.read_i32(
            if(seg.fileoff == 0 && seg.size != 0)
            {
                base_seg = seg;
               // alert(base_seg.addr)
            }
        }
        if(base_seg == null)
        {
            fail("base_seg");
        }
        var cache_slide = Sub(hdr, base_seg.addr);
        //alert("cache slide" + cache_slide + "vs" + memory.u32(hdr+0x38));
        //base_seg.addr = Add(base_seg.addr, cache_slide
        for(var i = 0; i < segs.length; ++i)
        {
            segs[i].addr = Add(segs[i].addr, cache_slide);
        }
        var libs =
        {
            "/usr/lib/system/libdyld.dylib":                                        ["_dlsym", "_dlopen"],
            "/usr/lib/system/libsystem_platform.dylib":                             ["__longjmp", "__platform_memmove"],
            "/usr/lib/system/libsystem_kernel.dylib":                               ["_mach_task_self_", "__kernelrpc_mach_vm_protect_trap"],
            "/usr/lib/system/libsystem_c.dylib":                                    ["_usleep"],
            "/System/Library/Frameworks/JavaScriptCore.framework/JavaScriptCore":   [
                "__ZN3JSC38taggedStartOfFixedExecutableMemoryPoolE",
                "__ZN3JSC36taggedEndOfFixedExecutableMemoryPoolE",
            ],
        };
        var opcodes;
        var opcode_libs;
            libs["/System/Library/Frameworks/JavaScriptCore.framework/JavaScriptCore"].push(
                "__ZN3JSC29jitWriteSeparateHeapsFunctionE"
            )
            opcodes = {
                // ldr x8, [sp] ; str x8, [x19] ; ldp x29, x30, [sp, #0x20] ; ldp x20, x19, [sp, #0x10] ; add sp, sp, #0x30 ; ret
                "ldrx8":       [0xf94003e8, 0xf9000268, 0xa9427bfd, 0xa9414ff4, 0x9100c3ff, 0xd65f03c0],
                // blr x21; ldp x29, x30, [sp, 0x30]; ldp x20, x19, [sp, 0x20]; ldp x22, x21, [sp, 0x10]; add sp, sp, 0x40; ret
                "dispatch":    [ 0xd63f02a0, 0xa9437bfd, 0xa9424ff4, 0xa94157f6, 0x910103ff, 0xd65f03c0 ],
                // mov x3, x22 ; mov x6, x27 ; mov x0, x24 ; mov x1, x19 ; mov x2, x23 ; ldr x4, [sp] ; blr x8
                "regloader":   [ 0xaa1603e3, 0xaa1b03e6, 0xaa1803e0, 0xaa1303e1, 0xaa1703e2, 0xf94003e4, 0xd63f0100 ],
                // ldp x29, x30, [sp, 0x60]; ldp x20, x19, [sp, 0x50]; ldp x22, x21, [sp, 0x40]; ldp x24, x23, [sp, 0x30];
                // ldp x26, x25, [sp, 0x20]; ldp x28, x27, [sp, 0x10]; add sp, sp, 0x70; ret
                "stackloader": [ 0xa9467bfd, 0xa9454ff4, 0xa94457f6, 0xa9435ff8, 0xa94267fa, 0xa9416ffc, 0x9101c3ff, 0xd65f03c0 ],
                // mov x4, x20 ; blr x8
                "movx4":       [ 0xaa1403e4, 0xd63f0100 ],
                // ldr x0, [x0] -- for debugging
                "ldrx0x0":     [ 0xf9400000 ],
            }
            opcode_libs = [
              //  "/usr/lib/libpthread.dylib",     // dispatch, stackloader
               // "/usr/lib/libc++.dylib",  // ldrx8, regloader, movx4, stackloader
                "/System/Library/Frameworks/ModelIO.framework/ModelIO",
                "/System/Library/Frameworks/CoreAudio.framework/CoreAudio",
            ];
        
        //print('lookin through cache');
       
        var syms = {};
        var gadgets = {};
      //  alert("dyld base address" + memory.u32(hdr + 0x20))
        var imgs  = Add(hdr, memory.u32(Add(hdr, 0x88)));
        //alert("imgs"+imgs)
        var nimgs = memory.u32(Add(hdr, 0x1c));
        //alert(nimgs + "n")
        for(var i = 0; i < nimgs; ++i)
        {
            var straddr = off2addr(segs, memory.u32(Add(imgs,i * 0x20 + 0x1c)));
            var fn = function(i,oz)
            {  if(oz) { return memory.read(Add(straddr, i), 0x10);
            }
                return memory.read(Add(straddr, i), 1)[0];
            };
           // alert(ab2str(fn(i,0)))
            var base = Add(memory.readInt64(Add(imgs, i * 0x20 + 0x10)), cache_slide);
           // alert(base + "vs" + memory.u32(Add(imgs,i * 0x20 + 0x10)))
            if(opcode_libs.some(lib => strcmp(fn, lib)))
            {
               //alert("here1")
                var ncmds = memory.u32(Add(base, 0x10));
                for(var j = 0, off = 0x20; j < ncmds; ++j)
                {
                // alert("here")
                    var cmd = memory.u32(Add(base, off));
                    if(cmd == 0x19 && strcmp(memory.read(Add(base, off + 0x10), 0x10), "__TEXT")) // LC_SEGMENT_64
                    {
                        var nsects = memory.u32(Add(base, off + 0x40));
                        for(var k = 0, let o = off + 0x48; k < nsects; ++k)
                        {
                            if(strcmp(memory.read(Add(base, o), 0x10), "__text"))
                            {
                                var keys = Object.keys(opcodes).filter(k=>!gadgets.hasOwnProperty[k])
                                if (keys.length == 0) break;
                                var match = {};
                                for(var z = 0; z < keys.length; ++z)
                                {
                                    match[keys[z]] = 0;
                                }
                                var addr = Add(memory.readInt64(Add(base, o + 0x20)), cache_slide)
                                var size = memory.u32(Add(base, o + 0x28))
                                //alert("here")
                                // Copy the entire __text region into a Uint32Array for faster processing.
                                // Previously you could map a Uint32Array over the data, but on i7+ devices
                                // this caused access violations.
                                // Instead we read the entire region and copy it into a Uint32Array. The
                                // memory.read primitive has a weird limitation where it's only able to read
                                // up to 4096 bytes. to get around this we'll read multiple times and combine
                                // them into one.
                                var allData = new Uint32Array(size / 4)
                                for (var r = 0; r < size; r += 4096) {
                                    // Check to ensure we don't read out of the region we want
                                    var qty = 4096
                                    if (size - r < qty) {
                                        qty = size - r
                                    }
                                    var data = memory.read(Add(addr, r), qty)
                                    // Data is an array of single bytes. This code takes four entries
                                    // and converts them into a single 32-bit integer. It then adds it
                                    // into the `allData` array at the given index
                                    for (var h = 0; h < qty; h += 4) {
                                        var fourBytes = b2u32(data.slice(h, h + 4))
                                        allData[(r + h) / 4] = fourBytes
                                    }
                                }
                                // Loop through the entire data map looking for each gadget we need
                                for (var f = 0; f < size && keys.length > 0; f++) {
                                    var op = allData[f]
                                    for (var z = 0; z < keys.length; z++) {
                                        var key = keys[z]
                                        var opcode = opcodes[key]
                                        if (op == opcode[match[key]]) {
                                            match[key]++
                                            if (match[key] == opcode.length) {
                                                gadgets[key] = Add(addr, (f - (opcode.length - 1)) * 4)
                                                keys.splice(z, 1)
                                                break
                                            }
                                        } else {
                                            match[key] = 0
                                        }
                                    }
                                }
                                break
                            }
                            o += 0x50;
                        }
                        break;
                    }
                    off += memory.u32(Add(base, off + 0x4));
                }
                continue;
            }
            var lookup = null;
            for(var k = Object.keys(libs), j = 0; j < k.length; ++j)
            {
                if(strcmp(fn, k[j]))
                {
                    lookup = libs[k[j]];
                    break;
                }
            }
            let mem3 = memory;
            if(lookup != null)
            {
                fsyms(mem3, base, segs, lookup, syms);
            }
        }
        if(!gadgets['dispatch'])
        {
            gadgets['dispatch'] = gadgets['altdispatch'];
        }
        if(!gadgets['stackloader'])
        {
            gadgets['stackloader'] = gadgets['altstackloader'];
        }
        delete opcodes['altdispatch'];
        delete opcodes['altstackloader'];
        var k = Object.values(libs).reduce(function(p,c){ c.forEach(function(e){ p.push(e) });return p; }, []);
        for(var i = 0; i < k.length; ++i)
        {
            var s = k[i];
            if(syms[s] == null)
            {
                fail(s);
            }
            syms[s] = Add(syms[s], cache_slide);
        }
        k = Object.keys(opcodes);
        for(var i = 0; i < k.length; ++i)
        {
            var s = k[i];
            if(gadgets[s] == null)
            {
                fail(s);
            }
        }
     alert("done! Exploitation complete stage1")*/
}"




Will this help you or not in see the mach-O format change but here's the current crash log 


"bootywarrior@Bootys-MacBook-Air ~ % /Users/bootywarrior/Downloads/ipsw_3.1.30_macOS_arm64/ipsw symbolicate /Users/bootywarrior/Desktop/web.ips /Users/bootywarrior/Downloads/dyld_shared_cache_arm64
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x98 pc=0x1037561c0]

goroutine 1 [running]:
github.com/blacktop/ipsw/cmd/ipsw/cmd.glob..func63(0x1042ef1c0, {0x14000253340, 0x2, 0x2})
	github.com/blacktop/ipsw/cmd/ipsw/cmd/symbolicate.go:154 +0x620
github.com/spf13/cobra.(*Command).execute(0x1042ef1c0, {0x14000253300, 0x2, 0x2})
	github.com/spf13/cobra@v1.3.0/command.go:856 +0x668
github.com/spf13/cobra.(*Command).ExecuteC(0x1042eea40)
	github.com/spf13/cobra@v1.3.0/command.go:974 +0x410
github.com/spf13/cobra.(*Command).Execute(...)
	github.com/spf13/cobra@v1.3.0/command.go:902
github.com/blacktop/ipsw/cmd/ipsw/cmd.Execute()
	github.com/blacktop/ipsw/cmd/ipsw/cmd/root.go:57 +0x30
main.main()
	github.com/blacktop/ipsw/cmd/ipsw/main.go:27 +0x20
bootywarrior@Bootys-MacBook-Air ~ % /Users/bootywarrior/Downloads/ipsw_3.1.31_macOS_arm64/ipsw symbolicate /Users/bootywarrior/Desktop/web.ips /Users/bootywarrior/Downloads/dyld_shared_cache_arm64
   ⨯ image WebCore not found in /Users/bootywarrior/Downloads/dyld_shared_cache_arm64
   ⨯ image WebCore not found in /Users/bootywarrior/Downloads/dyld_shared_cache_arm64
   ⨯ image WebCore not found in /Users/bootywarrior/Downloads/dyld_shared_cache_arm64
   ⨯ image WebCore not found in /Users/bootywarrior/Downloads/dyld_shared_cache_arm64
   ⨯ image WebCore not found in /Users/bootywarrior/Downloads/dyld_shared_cache_arm64
   ⨯ image WebCore not found in /Users/bootywarrior/Downloads/dyld_shared_cache_arm64
   ⨯ image WebCore not found in /Users/bootywarrior/Downloads/dyld_shared_cache_arm64
   ⨯ image WebCore not found in /Users/bootywarrior/Downloads/dyld_shared_cache_arm64
   ⨯ image WebCore not found in /Users/bootywarrior/Downloads/dyld_shared_cache_arm64
   ⨯ image WebCore not found in /Users/bootywarrior/Downloads/dyld_shared_cache_arm64
   ⨯ image WebKit not found in /Users/bootywarrior/Downloads/dyld_shared_cache_arm64
   ⨯ image WebKit not found in /Users/bootywarrior/Downloads/dyld_shared_cache_arm64
   ⨯ image WebKit not found in /Users/bootywarrior/Downloads/dyld_shared_cache_arm64
   ⨯ image WebKit not found in /Users/bootywarrior/Downloads/dyld_shared_cache_arm64
   ⨯ image JavaScriptCore not found in /Users/bootywarrior/Downloads/dyld_shared_cache_arm64
   ⨯ image JavaScriptCore not found in /Users/bootywarrior/Downloads/dyld_shared_cache_arm64
   ⨯ image CoreFoundation not found in /Users/bootywarrior/Downloads/dyld_shared_cache_arm64
   ⨯ image CoreFoundation not found in /Users/bootywarrior/Downloads/dyld_shared_cache_arm64
   ⨯ image CoreFoundation not found in /Users/bootywarrior/Downloads/dyld_shared_cache_arm64
   ⨯ image CoreFoundation not found in /Users/bootywarrior/Downloads/dyld_shared_cache_arm64
   ⨯ image CoreFoundation not found in /Users/bootywarrior/Downloads/dyld_shared_cache_arm64
   ⨯ image Foundation not found in /Users/bootywarrior/Downloads/dyld_shared_cache_arm64
   ⨯ image Foundation not found in /Users/bootywarrior/Downloads/dyld_shared_cache_arm64
   ⨯ image libxpc.dylib not found in /Users/bootywarrior/Downloads/dyld_shared_cache_arm64
   ⨯ image libxpc.dylib not found in /Users/bootywarrior/Downloads/dyld_shared_cache_arm64
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x98 pc=0x1009ee744]

goroutine 1 [running]:
github.com/blacktop/ipsw/cmd/ipsw/cmd.glob..func63(0x1015871c0, {0x140001e5340, 0x2, 0x2})
	github.com/blacktop/ipsw/cmd/ipsw/cmd/symbolicate.go:158 +0xd24
github.com/spf13/cobra.(*Command).execute(0x1015871c0, {0x140001e5300, 0x2, 0x2})
	github.com/spf13/cobra@v1.3.0/command.go:856 +0x668
github.com/spf13/cobra.(*Command).ExecuteC(0x101586a40)
	github.com/spf13/cobra@v1.3.0/command.go:974 +0x410
github.com/spf13/cobra.(*Command).Execute(...)
	github.com/spf13/cobra@v1.3.0/command.go:902
github.com/blacktop/ipsw/cmd/ipsw/cmd.Execute()
	github.com/blacktop/ipsw/cmd/ipsw/cmd/root.go:57 +0x30
main.main()
	github.com/blacktop/ipsw/cmd/ipsw/main.go:27 +0x20
"

**Comment by blacktop** [comment]
Just FYI, if this is your research you might not want to paste the backtrace here?

**Comment by turnerrocks1** [comment]
Oops. I meant to only drop the link wasn't paying attention about to go back and edit had the crash log in text editor!

**Comment by blacktop** [comment]
Also that looks like you are still using version `3.1.30` did you try it with version `3.1.31` ?

**Comment by blacktop** [comment]
oh nvm, I see you have both in your output

**Comment by turnerrocks1** [comment]
> Also that looks like you are still using version `3.1.30` did you try it with version `3.1.31` ?

oops my bad I have the same terminal open and haven't clear it next iteration I'll clear it !

**Comment by blacktop** [comment]
AAAH 😩 my apologies, my last fix was garbage.  This time I just fixed the dyld.Image func to return an error *(that I'll have to handle)* if it doesn't find the name.  It should be good now? 🤞

**Comment by turnerrocks1** [comment]
> AAAH 😩 my apologies, my last fix was garbage. This time I just fixed the dyld.Image func to return an error _(that I'll have to handle)_ if it doesn't find the name. It should be good now? 🤞

It's ok just waiting on the build to test it but I think its failing due to argument mismatch

**Comment by blacktop** [comment]
yeah, I missed one of them, should be up shortly

**Comment by turnerrocks1** [comment]
> yeah, I missed one of them, should be up shortly

ok thx!!!

**Comment by blacktop** [comment]
Let me know if setting the backtrace symbol too `"?"` when the image is NOT found makes sense?

**Comment by turnerrocks1** [comment]
> Let me know if setting the backtrace symbol too `"?"` when the image is NOT found makes sense?

actually I would implement that as it's kind of difficult to differentiate in it's current state but I have an update it finally works! 

Feature Idea is maybe detangle names? I know this is usually how the compiler and most decompilers range symbols is by having a preceeding __ZN7 symbol prefix but actually at runtime and even in the Javascriptcore/Webcore source code most function names and symbols aren't actually that way for example " 0: WebCore                     (slide=0x1e560000) 0x1a8306618 __ZN7WebCore15JSProgressEvent9prototypeERN3JSC2VMERNS_17JSDOMGlobalObjectE" which came from your ipsw symbolicator. In actually source code or IDA Pro decompiler pseudocode it would actually be Webcore::JSProgressEventprototype " VS Ghidra or hopper or even in your symbolicator instance "__ZN7WebCore15JSProgressEvent9prototypeERN3JSC2VMERNS_17JSDOMGlobalObjectE" also last but not least at least the symbolicator works and I can perform static analysis on the crash but I don't think the code you have wrangled the instance / object names correctly like the "ERN3JSC2VMERNS" seems like rambled garbage maybe your code has an uninitialized vector or overflow somewhere or maybe a glitch from this being an old shared cache? But I've read JSC and Webcore source code for the last two years consistently and there is nothing with that there, best regards and thanks a lot!

**Comment by blacktop** [comment]
`--demangle`

**Comment by blacktop** [comment]
Another think you can do is check out the syms like this:
```bash
> ipsw dyld a2s DSC ADDR
```
or
```bash
> ipsw dyld dump DSC ADDR
```
and if the dump is pointer to addresses
```bash
> ipsw dyld dump DSC ADDR -a
```

**Comment by turnerrocks1** [comment]
> --demangle

WOWWWWW!!! didn't know you already had that feature and it work's so perfectly !

**Comment by blacktop** [comment]
`ERN3JSC2VMERNS` is just part of the C++ mangled name that the `--demangle` flag will fix

**Comment by blacktop** [comment]
YA MAN! Don't EVER underestimate the POWER 💪 of `ipsw` 🤣

**Comment by blacktop** [comment]
Glad I could help 👍

**Comment by turnerrocks1** [comment]
> --demangle

last but not least like I said it works but I'm just now seeing this but it's unable to find the image WebContent. '⨯ image com.apple.WebKit.WebContent not found in cache'

**Comment by turnerrocks1** [comment]
> Another think you can do is check out the syms like this:
> 
> ```shell
> > ipsw dyld a2s DSC ADDR
> ```
> 
> or
> 
> ```shell
> > ipsw dyld dump DSC ADDR
> ```
> 
> and if the dump is pointer to addresses
> 
> ```shell
> > ipsw dyld dump DSC ADDR -a
> ```

the difference in the -a flag does what exactly ?

**Comment by blacktop** [comment]
Can you try `ipsw dyld info DSC --dylibs -V | grep Web`

**Comment by blacktop** [comment]
> > Another think you can do is check out the syms like this:
> > ```shell
> > > ipsw dyld a2s DSC ADDR
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > or
> > ```shell
> > > ipsw dyld dump DSC ADDR
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > and if the dump is pointer to addresses
> > ```shell
> > > ipsw dyld dump DSC ADDR -a
> > ```
> 
> the difference in the -a flag does what exactly ?

`-a` or `--addr` interprets and "slides" the data as pointers

**Comment by turnerrocks1** [comment]
> Can you try `ipsw dyld info DSC --dylibs -V | grep Web`

"bootywarrior@Bootys-MacBook-Air ~ % /Users/bootywarrior/Downloads/ipsw_3.1.32_macOS_arm64/ipsw dyld info /Users/bootywarrior/Downloads/dyld_shared_cache_arm64 --dylibs -V | grep Web 
   • Parsing Cache             cache=/Users/bootywarrior/Downloads/dyld_shared_cache_arm64
 201: 0x18930d000  B95B4899-D4C8-322A-B952-09BE1CFA5B4D  (7606.4.5.0.0)      /System/Library/PrivateFrameworks/WebCore.framework/Frameworks/libwebrtc.dylib
 203: 0x189821000  BC468784-F1B9-3705-A698-1B0270ADB90A  (7606.4.5.0.0)      /System/Library/PrivateFrameworks/WebCore.framework/WebCore
 204: 0x18b0c3000  72B67A3B-9870-30D3-B41C-6B75BA35F598  (7606.4.5.0.0)      /System/Library/PrivateFrameworks/WebKitLegacy.framework/WebKitLegacy
 321: 0x190847000  BA7E8E8E-F0FB-3FD1-B66B-59AF1B99F721  (7606.4.5.0.0)      /System/Library/Frameworks/WebKit.framework/WebKit
 330: 0x190f63000  F813D893-E7D9-3F9B-8ED8-364BF7FC71F1  (7606.4.5.0.0)      /System/Library/PrivateFrameworks/WebBookmarks.framework/WebBookmarks
 342: 0x192173000  1015F699-6012-3E5D-8F4D-4E2A167BE65F  (7606.4.5.0.0)      /System/Library/PrivateFrameworks/WebInspector.framework/WebInspector
 383: 0x19364f000  6C44E611-1451-3406-9102-819656EA5250  (180.5.1.0.0)       /System/Library/PrivateFrameworks/SharedWebCredentials.framework/SharedWebCredentials
 391: 0x193b64000  E142241A-4541-396A-93A8-4CC62F7D03BD  (7606.4.5.0.0)      /System/Library/PrivateFrameworks/WebUI.framework/WebUI
 825: 0x1a2218000  531D0C65-7D22-383B-917E-7B39C39A83FB  (2241.21.32.0.0)    /System/Library/AccessibilityBundles/WebCore.axbundle/WebCore
 826: 0x1a222f000  A171A9C2-2EE2-342D-B50D-47FA307A9852  (2241.21.32.0.0)    /System/Library/AccessibilityBundles/WebKit.axbundle/WebKit
 827: 0x1a2233000  A55B0C57-3E26-374C-8321-19D180A54B55  (2241.21.32.0.0)    /System/Library/AccessibilityBundles/WebKitLegacy.axbundle/WebKitLegacy
 828: 0x1a223c000  990853FC-4F23-3998-B586-EABF9291FD0A  (2241.21.32.0.0)    /System/Library/AccessibilityBundles/WebProcess.axbundle/WebProcess
 829: 0x1a2240000  0019BA88-C781-3104-8F51-8778BCECC36D  (2241.21.32.0.0)    /System/Library/AccessibilityBundles/WebProcessLoader.axbundle/WebProcessLoader
1137: 0x1a965a000  2115CE46-0BED-3AB0-8271-03FBAD252499  (7606.4.5.0.0)      /System/Library/PrivateFrameworks/WebApp.framework/WebApp
1138: 0x1a965f000  64E36EC0-6F19-35E9-839A-17EDD4ABCB68  (23.0.0.0.0)        /System/Library/PrivateFrameworks/WebContentAnalysis.framework/WebContentAnalysis
bootywarrior@Bootys-MacBook-Air ~ % 
"

**Comment by blacktop** [comment]
Another cool feature of the `symbolicate` command is the `--unslide` flag if you haven't used it yet.  It will "unslide" the crashlog making it easier to compare with IDA etc

**Comment by turnerrocks1** [comment]
> Another cool feature of the `symbolicate` command is the `--unslide` flag if you haven't used it yet. It will "unslide" the crashlog making it easier to compare with IDA etc

nice will check that out right now I was about to just load the shared cache with the slide ipsw says it has but that would be a better use too. !

**Comment by blacktop** [comment]
> > Can you try `ipsw dyld info DSC --dylibs -V | grep Web`
> 
> ```shell
> "bootywarrior@Bootys-MacBook-Air ~ % /Users/bootywarrior/Downloads/ipsw_3.1.32_macOS_arm64/ipsw dyld info /Users/bootywarrior/Downloads/dyld_shared_cache_arm64 --dylibs -V | grep Web 
>    • Parsing Cache             cache=/Users/bootywarrior/Downloads/dyld_shared_cache_arm64
>  201: 0x18930d000  B95B4899-D4C8-322A-B952-09BE1CFA5B4D  (7606.4.5.0.0)      /System/Library/PrivateFrameworks/WebCore.framework/Frameworks/libwebrtc.dylib
>  203: 0x189821000  BC468784-F1B9-3705-A698-1B0270ADB90A  (7606.4.5.0.0)      /System/Library/PrivateFrameworks/WebCore.framework/WebCore
>  204: 0x18b0c3000  72B67A3B-9870-30D3-B41C-6B75BA35F598  (7606.4.5.0.0)      /System/Library/PrivateFrameworks/WebKitLegacy.framework/WebKitLegacy
>  321: 0x190847000  BA7E8E8E-F0FB-3FD1-B66B-59AF1B99F721  (7606.4.5.0.0)      /System/Library/Frameworks/WebKit.framework/WebKit
>  330: 0x190f63000  F813D893-E7D9-3F9B-8ED8-364BF7FC71F1  (7606.4.5.0.0)      /System/Library/PrivateFrameworks/WebBookmarks.framework/WebBookmarks
>  342: 0x192173000  1015F699-6012-3E5D-8F4D-4E2A167BE65F  (7606.4.5.0.0)      /System/Library/PrivateFrameworks/WebInspector.framework/WebInspector
>  383: 0x19364f000  6C44E611-1451-3406-9102-819656EA5250  (180.5.1.0.0)       /System/Library/PrivateFrameworks/SharedWebCredentials.framework/SharedWebCredentials
>  391: 0x193b64000  E142241A-4541-396A-93A8-4CC62F7D03BD  (7606.4.5.0.0)      /System/Library/PrivateFrameworks/WebUI.framework/WebUI
>  825: 0x1a2218000  531D0C65-7D22-383B-917E-7B39C39A83FB  (2241.21.32.0.0)    /System/Library/AccessibilityBundles/WebCore.axbundle/WebCore
>  826: 0x1a222f000  A171A9C2-2EE2-342D-B50D-47FA307A9852  (2241.21.32.0.0)    /System/Library/AccessibilityBundles/WebKit.axbundle/WebKit
>  827: 0x1a2233000  A55B0C57-3E26-374C-8321-19D180A54B55  (2241.21.32.0.0)    /System/Library/AccessibilityBundles/WebKitLegacy.axbundle/WebKitLegacy
>  828: 0x1a223c000  990853FC-4F23-3998-B586-EABF9291FD0A  (2241.21.32.0.0)    /System/Library/AccessibilityBundles/WebProcess.axbundle/WebProcess
>  829: 0x1a2240000  0019BA88-C781-3104-8F51-8778BCECC36D  (2241.21.32.0.0)    /System/Library/AccessibilityBundles/WebProcessLoader.axbundle/WebProcessLoader
> 1137: 0x1a965a000  2115CE46-0BED-3AB0-8271-03FBAD252499  (7606.4.5.0.0)      /System/Library/PrivateFrameworks/WebApp.framework/WebApp
> 1138: 0x1a965f000  64E36EC0-6F19-35E9-839A-17EDD4ABCB68  (23.0.0.0.0)        /System/Library/PrivateFrameworks/WebContentAnalysis.framework/WebContentAnalysis
> bootywarrior@Bootys-MacBook-Air ~ % 
> ```

Yeah man, I'm not seeing `WebContent` in there.

**Comment by turnerrocks1** [comment]
Weird thing is web content is the process itself that safari register under and when code execution is gain that's the context that it falls under is com.apple.webkit.webcontent so I don't really know lol

**Comment by turnerrocks1** [comment]
hey I found out why! SO after opening the jailbroken device and installing files to see why

**Comment by blacktop** [comment]
> > Another cool feature of the `symbolicate` command is the `--unslide` flag if you haven't used it yet. It will "unslide" the crashlog making it easier to compare with IDA etc
> 
> nice will check that out right now I was about to just load the shared cache with the slide ipsw says it has but that would be a better use too. !

LOL YEAH, I really like the `symbolicate` command because I put a lot of thought into it and added those nice to have features, in my opinion almost making it better than having to wait for XCode to do it (if that is even a possibility for people)

I guess it's kinda niche, but it's still one of my favorite commands ❤️

**Comment by blacktop** [comment]
`WebContent` is probably a userspace process and not in the cache is what I would guess

**Comment by turnerrocks1** [comment]
I went on iOS 12.1.4 and loaded up filza and went to "/System/Libary/PrivateFrameworks/Webkit.framework/XPCServices" and there were three folders "com.apple.WebKit.Networking.xpc" and "com.apple.WebKit.Storage.xpc" and "com.apple.WebKit.Webcontent.xpc" and in each folder listed there is a corresponding executable with -rwxr-xr-x permissions why it's not in the cache I don't know but then again it has to be a cache image to be in the crash log ?

**Comment by turnerrocks1** [comment]
> > > Another cool feature of the `symbolicate` command is the `--unslide` flag if you haven't used it yet. It will "unslide" the crashlog making it easier to compare with IDA etc
> > 
> > 
> > nice will check that out right now I was about to just load the shared cache with the slide ipsw says it has but that would be a better use too. !
> 
> LOL YEAH, I really like the `symbolicate` command because I put a lot of thought into it and added those nice to have features, in my opinion almost making it better than having to wait for XCode to do it (if that is even a possibility for people)
> 
> I guess it's kinda niche, but it's still one of my favorite commands ❤️

yessss it really helped me out because I just recently got a MacBook randomly yesterday so Xcode wasn't an option for me and limited me to javascript exploitation so yes this feature not only helps me but I'm pretty Sure every person that is unfortunate to own an up to date MacBook!

**Comment by blacktop** [comment]
No each Safari tab spawns a `WebContent` process and the shared_cache gets mapped into ever process on the device.  So that might be a weakness where XCode is better because it can probably as the phone about that process and symbolicate it's symbols etc.  Where as `ipsw` would have to also parse that MachO on the phone.

**Comment by blacktop** [comment]
That might be overkill though as I kind of envisioned this command as the "someone gives you a random crashdump to analyze and you have no access to the device to symbolicate it with XCode.

**Comment by turnerrocks1** [comment]
> No each Safari tab spawns a `WebContent` process and the shared_cache gets mapped into ever process on the device. So that might be a weakness where XCode is better because it can probably ask the phone about that process (macho) and symbolicate its symbols etc.
> 
> Where as `ipsw` would have to also parse that MachO on the phone.

nice explanation I understand now so basically Xcode has a head over since it can ask for symbolication at runtime vs ipsw being limited to post mortum of a crash log

**Comment by blacktop** [comment]
One crappy way you could get the symbol for the `WebContent` is to pull it off the phone and run:

```bash
> ipsw macho info WebContent --symbols
``` 
but then you'd have to apply the slide to them all which is gross

**Comment by blacktop** [comment]
I'm working on a branch now where I'm completely revamping my `diassembler` too where I plan to support `slid` disass.  So when that's done you could also pull the WebContent binary down and run
```bash
> ipsw disass WebContent -a ADDR -s SLIDE
```
but that won't work until the branch is done and merged
